jni/src/faiss_index_service.cpp

```diff
@@ -160,5 +160,119 @@ void BinaryIndexService::createIndex(
     faissMethods->writeIndexBinary(idMap.get(), indexPath.c_str());
 }
 
+ByteIndexService::ByteIndexService(std::unique_ptr<FaissMethods> faissMethods) : IndexService(std::move(faissMethods)) {}
+
+    std::unique_ptr <faiss::Index> ByteIndexService::validate(
+            knn_jni::JNIUtilInterface *jniUtil,
+            JNIEnv *env,
+            int vectorSize,
+            faiss::MetricType metric,
+            std::string indexDescription,
+            int dim,
+            int numIds,
+            int threadCount,
+            std::unordered_map <std::string, jobject> parameters
+    ) {
+        if (vectorSize == 0) {
+            throw std::runtime_error("Number of vectors cannot be 0");
+        }
+
+        // The number of vectors can be int here because a lucene segment number of total docs never crosses INT_MAX value
+        int numVectors = (int) (vectorSize / (uint64_t) dim);
+        if (numVectors == 0) {
+            throw std::runtime_error("Number of vectors cannot be 0");
+        }
+
+        if (numIds != numVectors) {
+            throw std::runtime_error("Number of IDs does not match number of vectors");
+        }
+
+        std::unique_ptr <faiss::Index> indexWriter(faissMethods->indexFactory(dim, indexDescription.c_str(), metric));
+
+        // Set thread count if it is passed in as a parameter. Setting this variable will only impact the current thread
+        if (threadCount != 0) {
+            omp_set_num_threads(threadCount);
+        }
+
+        // Add extra parameters that cant be configured with the index factory
+        SetExtraParameters<faiss::Index, faiss::IndexIVF, faiss::IndexHNSW>(jniUtil, env, parameters,
+                                                                            indexWriter.get());
+
+        // Check that the index does not need to be trained
+        if (!indexWriter->is_trained) {
+            throw std::runtime_error("Index is not trained");
+        }
+
+        return indexWriter;
+
+    }
+
+    void ByteIndexService::addVectorsToIndex(
+            std::unique_ptr <faiss::Index> indexWriter,
+            std::vector <int8_t> *inputVectors,
+            int dim,
+            std::vector <int64_t> ids,
+            std::string indexPath
+    ) {
+        std::unique_ptr <faiss::IndexIDMap> idMap(faissMethods->indexIdMap(indexWriter.get()));
+        int numVectors = (int) (inputVectors->size() / (uint64_t) dim);
+
+        int batchSize = 1000;
```

Comment:
why 1000?

Reply:
I ran some tests comparing a batch size of 1 vs 1000, I could see good difference in the forcemerge time where a batch size of 1000 performs faster by consuming little bit of extra memory